### DISCUSSION TITLE
[General] Cannot use web3.js version 4.x with zksync

### DISCUSSION THREAD

**Discussion by stnong** [zkSync-Community-Hub/zksync-developers#105] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

We are trying to call a write function on a contract using web3.js version 4.x

### What is the specific issue or error you're encountering?

When calling a write function on a contract we are getting this error: "InvalidResponseError: Returned error: Method not found". When using an older version of web3js or the zksync sdk or ethers it works fine. But we have to use the latest version of web3js because of proxy support and other issues.

### Can you share the error messages or logs you're receiving, if any?

InvalidResponseError: Returned error: Method not found

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

"web3": "^4.0.3",
"https-proxy-agent": "^5.0.1",

### Can you provide the relevant portions of your contract code where the issue is occurring?

`      const gas = await this.getWeb3Contract()
        .methods.addToPaymasterWhitelist(tokenAddress)
        .estimateGas({ from: getWalletAddress() });
      await this.getWeb3Contract()
        .methods.addToPaymasterWhitelist(tokenAddress)
        .send({ from: getWalletAddress(), gas: gas.toString() });`

### Have you tried to isolate the problem, and if so, what were the results?

.

### What steps have you already taken to try to resolve the issue?

- Create a raw tx, sign it and send it that way instead of through a contract object.
- Made sure its not an abi or contract problem by trying with ethers or zksync sdk.

**Comment by uF4No** [comment]
Hi @stnong , thanks for letting us know. 
We've not fully tested the latest version of web3.js. We recommend using zksync-web3, which was built on top of ethers v5 as that's the only one fully supported (apart from [the other SDKs](https://era.zksync.io/docs/tools/))

**Reply by stnong** [reply]
Hi @uF4No, problem with zksync-web3/ethers v5 is that it doesnt have support for https proxy agent which is necessary for us.

**Reply by MexicanAce** [reply, marked as answer]
Hi @stnong 👋 

Ethers support for a custom http.agent is [being actively discussed here in the ethers.js repo](https://github.com/ethers-io/ethers.js/discussions/2775). Also, unfortunately there's no current plan for support of `web3.js`.

**Reply by stnong** [reply]
Hi @MexicanAce,

it looks like ethers v6 already has custom http agent support. But zksync-web3 runs on ethers v5. Any plans when ethers v6 support will come out for zksync?

**Reply by MexicanAce** [reply]
Yes! We're actively working on supporting ethers v6 and will have further announcements about this soon on our [@zkSyncDevs](https://twitter.com/zkSyncDevs) Twitter account.

**Comment by grantimagen** [comment]
The "Method not found" error typically indicates that the method being called on the contract does not exist or is not available. Ensure that the method addToPaymasterWhitelist is correctly defined in your contract and matches the method signature being called in your JavaScript code. Additionally, verify that the contract ABI used in your JavaScript code matches the ABI of the deployed contract. If the issue persists, consider consulting the documentation or community resources for the specific version of Web3.js you are using to ensure compatibility with the contract's methods.

**Comment by cahillnick** [comment]
It seems that the method "addToPaymasterWhitelist" is not recognized by the contract. 
 Make sure the method name and parameters are correct, and that the contract ABI is properly configured.

**Comment by shelleyalexander** [comment]
The error "Method not found" typically indicates that the method being called on the contract does not exist or cannot be found. Ensure that the method name and parameters are correct and match the contract's ABI. Also, make sure that the contract address and ABI are correctly initialized in your web3.js setup.

**Comment by marcellusjustin** [comment]
The user is encountering an "InvalidResponseError: Returned error: Method not found" error when calling a write function on a contract using web3.js version 4.x. They have tried using an older version of web3js or the zksync sdk or ethers, which worked fine, but they need to use the latest version of web3js. They have also tried creating a raw tx, signing it, and sending it that way instead of through a contract object. The relevant code involves estimating gas and sending a transaction to addToPaymasterWhitelist function.

**Comment by mooralfeapur** [comment]
The user is encountering an "InvalidResponseError: Returned error: Method not found" when calling a write function on a contract using web3.js version 4.x on the zkSync Testnet. The user has not made any recent changes to the contract before encountering this issue. The contract interacts with the "web3" and "https-proxy-agent" external libraries. The user has provided relevant portions of their contract code where the issue is occurring. The user has tried to isolate the problem by creating a raw tx, signing it, and sending it that way instead of through a contract object, and by trying with ethers or zksync sdk. No repo link is provided.
